polkadot/xcm/xcm-builder/src/routing.rs

```diff
@@ -139,3 +139,69 @@ impl EnsureDelivery for Tuple {
 		(None, None)
 	}
 }
+
+/// A wrapper router that attempts to *encode* and *decode* passed XCM `message` to ensure that the
+/// receiving side will be able to decode, at least with the same XCM version.
+///
+/// This is designed to be at the top-level of any routers which do the real delivery. While other
+/// routers can manipulate the `message`, we cannot access the final XCM due to the generic
+/// `Inner::Ticket`. Therefore, this router aims to validate at least the passed `message`.
+///
+/// NOTE: For use in mock runtimes which don't have the DMP/UMP/HRMP XCM validations.
```

Comment:
Why would this only be used for testing? I think it is more important for produciton deployment to prevent invalid messages from ever being sent.

Reply:
well, the thing is that when we wrap `EnsureDecodableXcm<XcmpQueue>` it will be able to check message which **is passed** to the `XcmpQueue`, which could be ok, but we have other routers which internally can modify message and possibly break validity.  

We could just wrap `EnsureDecodableXcm<XcmpQueue>`, but we pass `XcmpQueue` on several places:
```
SovereignPaidRemoteExporter<bridging::EthereumNetworkExportTable, XcmpQueue, UniversalLocation>
...
type ToBridgeHubSender = XcmpQueue;
...
```

Yes, we could wrap "everywhere", but it can be easily missed.
So, I think, if we want to be pretty sure, this validation should be as a part of `XcmpQueue` / `ParentAsUmp` / `ChildParachainRouter` internal implementation, and I think that `SendXcm::validate` is the best place (as you put it for `XcmpQueue`) and also there is no need to do double-check in `SendXcm::deliver`

Reply:
> So, I think, if we want to be pretty sure, this validation should be as a part of XcmpQueue / ParentAsUmp / ChildParachainRouter internal implementation

Yea 👍 did not want to suggest it since i though that Router would be better, but if it always needs to wrapped then just put it directly in the pallet.

Reply:
`validate_xcm_nesting` changes are moved to the separate PR: https://github.com/paritytech/polkadot-sdk/pull/4236